lib/utils/README.md

```diff
@@ -2,19 +2,375 @@
+### Motivation
+
+FlexFlow's codebase makes heavy use of "plain old data"[^2] types[^1] (referred to these as _product types_ in the rest of this document) such as the following:
+```cpp
+struct Person {
+  std::string first_name;
+  std::string last_name;
+  int age;
+};
+```
+However, this standard implementation defines a set of behaviors that we (i.e., the FlexFlow developers) find undesirable.
+These are as follows:
+
+1. Default constructibility: for many product types default constructibility can make code bug-prone. For example, let us consider the following valid code:
+```cpp
+Person some_function() {
+  Person p;
+  if (...) {
+    p = {"donald", "knuth", 85};
+  }
+  return p;
+}
+```
+If the `if` branch is not taken, we will return a `Person` with nonsensical values, and there do not exist any values that naturally form a default.
+For example, we could initalize the values as follows
+```cpp
+struct Person {
+  std::string first_name; // initializes to ""
+  std::string last_name; // initializes to ""
+  int age = 0;
+}
+```
+But this is a completely useless value, and if it shows up anywhere in our code it's probably a bug, since a nameless, age 0 person is probably not a helpful value to have.
+
+2. Even if we disable default constructibility, the ability to perform aggregate initialization of a subset of the `struct`'s fields is still undesirable. For example,
+```cpp
+struct Person {
+  Person() = delete;
+
+  std::string first_name;
+  std::string last_name;
+  int age = 0;
+};
+
+Person p{"donald", "knuth"};
+```
+will compile just fine, but will silently use a nonsensical value of `age`. Even worse, if we were to add an additional field `is_male`
 ```cpp
+struct Person {
+  Person() = delete;
+
+  std::string first_name;
+  std::string last_name;
+  int age = 0;
+  bool is_male = false;
+};
+
+Person p{"donald", "knuth", 85};
+```
+would compile without any errors, but is (as of writing this) incorrect.
+
+3. For product types, `operator==` and `operator!=` are trivial, but still have to be written and maintained, and can easily lead to bugs. For example, 
+```
+struct Person {
+  Person() = delete;
+  Person(std::string const &first_name, 
+         std::string const &last_name, 
+         int age)
+    : first_name(first_name),
+      last_name(last_name),
+      age(age),
+    { }
+
+  friend bool operator==(Person const &lhs, Person const &rhs) {
+    return lhs.first_name == rhs.first_name 
+      && lhs.last_name == rhs.last_name
+      && lhs.age == rhs.age;
+  }
+
+  friend bool operator!=(Person const &lhs, Person const &rhs) {
+    return lhs.first_name != rhs.first_name 
+      || lhs.last_name != rhs.last_name
+      || lhs.age != rhs.age;
+  }
 
+  std::string first_name;
+  std::string last_name;
+  int age;
+};
 ```
+If we take the previous example of adding an additional `is_male` field to `Person`, it can be easy to miss a location, leading to incorrectness. 
```

Comment:
Good examples, but maybe reword to " it can be easy to forget to update the operators."